CenterForOpenScience/osf.io
# [WIP] Normalize Education and Employment models/API endpoints #PLAT-1285

## Purpose

Education and employment were up til now stored on the user model, as a list of dictionaries. This PR pulls off that info into their own Education and Employment models that have a FK back to the user model. It also adds new education and employment API endpoints.

## Changes

- Add and populate Education and Employment models
- Add education and employment API endpoints
	- Includes the ability to update ordering using JSONAPI Relationship serializers
- Add new jsonschemas for validating fields
- Update the frontend to use the new API endpoints for updating/changing education and employment
- Lots of tests

## TODO
- Fully removing `jobs` and `schools` usage and references around the code
- Make sure merging users works as expected
- API versioning
	- from the ticket: "For older versions of the API, just include blank objects on the user serialization rather than removing the field entirely."
- Update developer API documentation
- Fix validation that is mistakenly allowing both a start and end time along with ongoing. There's a failing test for this.
- General polishing and sanity checking

## QA Notes

This is mostly backend related, but does update the frontend code to use the new models! This also involves a migration of data so should include some prep that you have some users with multiple education and employment before this migration is run.

Pre-migration:
- Have a user with multiple education and employment entries 
	- Make sure to note their order

Post-migration:
- Make sure that user has all their entries and in the correct order
- Make sure you can add new entries
- Make sure you can drag to re-order your entries on the frontent
- Make sure entries stick around and are in the correct order on page refresh

## Documentation

To be added!

## Side Effects

Shouldn't be many no!

## Ticket

https://openscience.atlassian.net/browse/PLAT-1285
